File: Vol.02(1100-1199)/#1196 Fast Food/p1196.cpp
// 2939471 2008-06-01 23:57:15 Accepted 1196 C++ 00:00.02 1024K 呆滞的慢板 

// DP !!! O(n^3)
// DP[k][n]: 表示从左手边起的 n 个站时，有 k 个快餐店的结果
// 另外要打一个表: DIS[i][j] 表示在第 i 个到第 j 个站放置 1 个快餐店的花费
// 然后有: DP[k][n] = min( DP[k-1][t] + DIS[t][n] ),  ( k <= t <= n )

#include <iostream>
using namespace std;

int A[200], N,  K,  T = 0;
int DP[31][201],    DIS[200][200];

int main() {
    
    while( cin >> N >> K && N ) {
        
        for( int i = 0; i < N; ++i )
            cin >> A[i];
        
        for( int i = 0; i < N; ++i ) {
            for( int j = i; j < N; ++j ) {
                int x = i,  y = j;
                DIS[i][j] = 0;
                while( x < y )
                    DIS[i][j] += ( A[y--] - A[x++] );
            }
        }

    #ifndef ONLINE_JUDGE
    for( int i = 0; i < N; ++i ) {
        for( int j = 0; j < N; ++j )
            printf( "%3d", DIS[i][j] );
        puts( "" );
    }
    #endif

        for( int i = 1; i <= N; ++i )
            DP[1][i] = DIS[0][i - 1];
            
        for( int i = 2; i <= K; ++i ) {
            for( int j = i; j <= N; ++j ) {
                DP[i][j] = 0x7FFFFFFF;
                for( int k = i - 1; k < j; ++k ) {
                    if( DP[i-1][k] + DIS[k][j-1] < DP[i][j] )
                        DP[i][j] = DP[i-1][k] + DIS[k][j-1]; 
                }
            }
        }

    #ifndef ONLINE_JUDGE
    puts( "" );
    for( int i = 0; i <= K; ++i ) {
        for( int j = 0; j <= N; ++j )
            printf( "%3d", DP[i][j] );
        puts( "" );
    }
    #endif

        cout << "Chain " << ++T << endl;
        cout << "Total distance sum = " << DP[K][N] << endl << endl;
        
    }

}

File: 范例/Accpt/11XX/p1196.pas
PROGRAM p1196;

CONST
        maxlongint=999999;

VAR
        n,k,i,step                      :Integer;
        num                             :array[0..301]of Longint;
        min                             :array[0..31,0..301]of Longint;
        dis                             :array[0..301,0..301]of Longint;

PROCEDURE Main;
var
        i,j,p                           :Integer;
        sum,ans                         :Longint;
begin
num[0]:=-maxlongint;
num[n+1]:=maxlongint;
for i:=1 to n+1 do min[0,i]:=maxlongint;
min[0,0]:=0;
for i:=0 to n do
        for j:=i+1 to n+1 do
                begin
                dis[i,j]:=0;
                for p:=i+1 to j-1 do
                        if abs(num[p]-num[i])>abs(num[j]-num[p]) then inc(dis[i,j],abs(num[j]-num[p]))
                                else inc(dis[i,j],abs(num[p]-num[i]));
                end;

for i:=1 to k do
        for j:=1 to n+1 do
                begin
                if i=j then
                        begin
                        min[i,j]:=0;
                        continue;
                        end;
                min[i,j]:=maxlongint;
                for p:=j-1 downto 0 do
                        begin
                        sum:=0;
                        sum:=dis[p,j];
                        if min[i-1,p]+sum<min[i,j] then min[i,j]:=min[i-1,p]+sum;
                        end;
                end;

ans:=maxlongint;
for i:=n downto 1 do
        begin
        sum:=0;
        for j:=i+1 to n do inc(sum,abs(num[j]-num[i]));
        if sum+min[k,i]<ans then ans:=min[k,i]+sum;
        end;
writeln('Chain ',step);
write('Total distance sum = ');
writeln(ans);
writeln;
end;

BEGIN
step:=0;
while true do
        begin
        inc(step);
        readln(n,k);
        if (n=0)and(k=0) then break;
        for i:=1 to n do readln(num[i]);
        Main;
        end;
END.